frontend/src/services/HTTPHelper.ts

```diff
@@ -0,0 +1,368 @@
+      const value = kvp.slice(indexOfColon + 1).trim();
+
+      if (key.length > 0 && value.length > 0) {
+        result[key] = value;
+      }
+    }
+  }
+
+  return result;
+}
+
+/**
+ * Builds the query service options.
+ * @param options The partial options that very specified externally.
+ * @returns The built options.
+ */
+function buildOptions(options: Partial<HTTPHelperOptions>): HTTPHelperOptions {
+  let { baseUri } = options;
+
+  // As per the spec of HTTPHelperOptions, we use the base URI of the source the frontend
+  // was loaded from, if no base uri was specified in the options.
+  if (!baseUri) {
+    const { location } = window;
+    baseUri = `${location.protocol}//${location.host}/${
+      location.pathname.split('/')[1]
+    }`;
+  }
+
+  return { baseUri };
+}
+
+@injectable()
+export default class HTTPHelper {
+  private readonly options: HTTPHelperOptions;
+
+  public constructor(options: Partial<HTTPHelperOptions> = {}) {
+    this.options = buildOptions(options);
+  }
+
+  public tryGet<TResult>(
+    url: string | URL,
+    request: HTTPGETRequest,
+    cancellation?: CancellationToken
+  ): Promise<HTTPResponse<TResult>> {
+    return this.executeRequest(url, request, 'get', undefined, cancellation);
+  }
+
+  public tryPost<TResult>(
+    url: string | URL,
+    request: HTTPPOSTRequest,
+    cancellation?: CancellationToken
+  ): Promise<HTTPResponse<TResult>> {
+    return this.executeRequest(
+      url,
+      request,
+      'post',
+      request.body,
+      cancellation
+    );
+  }
+
+  private executeRequest<TResult>(
+    url: string | URL,
+    request: HTTPRequest,
+    method: 'post' | 'get' | 'head' | 'put' | 'delete' | 'patch',
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
+    body?: any,
+    cancellation?: CancellationToken
+  ): Promise<HTTPResponse<TResult>> {
+    // Pre-process arguments, extract the header collection and set a default content type header, that can be overridden by the caller.
+    const headers: HTTPHeaderCollection = {
+      'Content-Type': 'application/json',
+      ...request.headers,
+    };
+
+    const parsedURL = this.appendQuery(url, request.query);
+
+    return new Promise<HTTPResponse<TResult>>((resolve, reject) => {
+      // If the cancellation token is already canceled, we can reject right away.
+      if (cancellation?.isCancellationRequested) {
+        reject(new CancellationError());
+      }
+
+      const httpClient = new XMLHttpRequest();
+      httpClient.open(method, parsedURL.href, true);
+
+      // Process the request-headers
+      // Walk over all attributes of the 'headers' object and set the key-value pairs as request header.
+      const keys = Object.keys(headers);
+      for (let i = 0; i < keys.length; i += 1) {
+        const key = keys[i];
+        const value = headers[key];
+
+        if (typeof value === 'string') {
```

Comment:
why don't we define `HTTPHeaderCollection` as
```ts
interface HTTPHeaderCollection {
  [key: string]: string; // removed  "| undefined"
}
```
Then this is tested at compile time

Reply:
That's because we don't have a value for any key. The type would not be correct, as typescript assumes that we can now pass an arbitrary string as key and get a string back. Not for all keys are there values however.